#165415819 Setup postgresql database

Nerocodes opened the pull request:
### What does this PR do?
>Adds postgresql database
>Modifies all endpoints to use database
>Modifies tests to use database

### Description of Task to be completed?
Setup postgresql database

### How should this be manually tested?
#### Locally
>navigate to [repo url](https://github.com/Nerocodes/banka) 
>Clone repo
>```` npm install ````
>```` npm start ````
#### Online
>view [documentation](https://documenter.getpostman.com/view/6692772/S1ENyyf9#0d24455e-8a99-4123-8990-b008622679d9)

### Any background context you want to provide?
NA

### What are the relevant pivotal tracker stories?
[#165415819](https://www.pivotaltracker.com/story/show/165415819)

### Screenshots (if appropriate)
NA

### Questions:
NA

Nerocodes committed:
ch(database): setup postgres database
modified all endpoint to use database
modified tests to use database
[Delivers #165415819]

Nerocodes committed:
bg(travis): check database travis is using

Nerocodes committed:
bg(travis): console error

Nerocodes committed:
bg(travis): fixed database error
[Finishes #165467373]

Nerocodes committed:
bg(travis): console log error

Nerocodes committed:
bg(travis): fixed database error
[Delivers #165467373]

Nerocodes committed:
bg(travis): pass tests

ucheg6 commented on api/package.json:
```diff
@@ -7,10 +7,19 @@
+    "singleTest": "set NODE_ENV='TEST' && nyc mocha --timeout=20000 --require @babel/register ./src/test/account.test.js --exit || true",
+    "coverage": "nyc npm test && nyc report --reporter=text-lcov | coveralls",
+    "migrateUser": "babel-node ./src/database/migrations/user createUserTable",
+    "migrateAccount": "babel-node ./src/database/migrations/account createAccountTable",
```
why are you migrating the tables differently?

Nerocodes replied:
I've merged it into one file now.

ucheg6 commented on api/src/controllers/account.controller.js:
```diff
@@ -42,33 +42,33 @@ const AccountController = {
-    });
-  },
+  // fetchAllAccounts(req, res) {
+  //   const allAccounts = AccountService.fetchAllAccounts(req);
```
Please no commented out codes

Nerocodes replied:
Removed

ucheg6 commented on api/src/controllers/account.controller.js:
```diff
@@ -2,8 +2,8 @@ import AccountService from '../services/account.service';
+    const createdAccount = await AccountService.createAccount(req.userId, req.body.type);
     if (createdAccount.error) {
       return res.json({
         status: 401,
```
you have errors without error messages. Please add descriptive error messages

Nerocodes replied:
There are error messages. The error messages are coming from the services. That's how it is for the rest to

Nerocodes replied:
I send the error messages as an object to the controller

ucheg6 commented on api/src/controllers/user.controller.js:
```diff
@@ -1,51 +1,81 @@
       newUser.type = 'client';
       newUser.isAdmin = false;
     }
+    if (!newUser.firstName) {
```
can this be done with express validator or joi and it should be on the route. Write a validation middleware and call it on the route and not on the controller.

Nerocodes replied:
Done. I used joi

ucheg6 commented on api/src/controllers/user.controller.js:
```diff
@@ -1,51 +1,81 @@
-    });
-    createdUser.token = token;
+    const createdUser = await UserService.addUser(newUser);
     Object.defineProperty(createdUser, 'password', {
```
This is unnecessary as you can decide the data to be returned from the query

Nerocodes replied:
I've refactored this now.

ucheg6 commented on api/src/controllers/user.controller.js:
```diff
@@ -1,51 +1,81 @@
+    if (!oldUser.email) {
+      return res.json({
+        status: 400,
+        error: 'Email field is required',
```
same here. write a validator middleware and call it on the route

Nerocodes replied:
Done. I used joi

ucheg6 commented on api/src/controllers/user.controller.js:
```diff
@@ -63,24 +93,20 @@ const UserController = {
-    }).status(200);
-  },
+  // getSingleUser(req, res) {
+  //   const { id } = req.params;
```
No comments!!!!!!!!!!!!!!!!!!!!!!!!!!!!

Nerocodes replied:
Removed

ucheg6 commented on api/src/database/database.js:
```diff
@@ -0,0 +1,29 @@
+}
+
+if (!process.env.NODE_ENV) {
+  DATABASE_URL = 'postgres://ysfjltsj:nOu1tZuIJf0D5K6oehu26V44fYTZmEqu@isilo.db.elephantsql.com:5432/ysfjltsj';
```
can't the DATABASE_URL's value come from .env?

Nerocodes replied:
I've added the environment variables to travis now.

ucheg6 commented on api/src/services/account.service.js:
```diff
@@ -1,94 +1,132 @@
+  //     const newAccount = new Account();
+  //     newAccount.id = singleAccount.id;
+  //     newAccount.accountNumber = singleAccount.accountNumber;
+  //     newAccount.createdOn = singleAccount.createdOn;
```
🙅‍♂️ 🙅‍♂️ .... REMOVE!!!!

Nerocodes replied:
Removed

ucheg6 commented on api/src/services/account.service.js:
```diff
@@ -1,94 +1,132 @@
-      .find(singleAccount => singleAccount.accountNumber == accountNumber);
-    return account || {};
-  },
+  // getAnAccount(accountNumber) {
```
comments!!!!

Nerocodes replied:
Removed

Nerocodes committed:
refactor(api): removed commented codes
merged migrations to one file
merged seeders to one file
removed database url string
modified validation helper

Nerocodes committed:
ch(heroku): updated build folder

Nerocodes committed:
ch(github): updated readme

Nerocodes committed:
bg(github): fixed table markdown

Nerocodes committed:
bg(github): fixed api markdown
